plugins/regex_revalidate/regex_revalidate.cc

```diff
@@ -0,0 +1,729 @@
+    date_loc = TSMimeHdrFieldFind(buf, hdr_loc, TS_MIME_FIELD_DATE, TS_MIME_LEN_DATE);
+    if (nullptr != date_loc) {
+      date = TSMimeHdrFieldValueDateGet(buf, hdr_loc, date_loc);
+      TSHandleMLocRelease(buf, hdr_loc, date_loc);
+    }
+    TSHandleMLocRelease(buf, nullptr, hdr_loc);
+  }
+
+  return date;
+}
+
+int
+main_handler(TSCont cont, TSEvent event, void *edata)
+{
+  TSHttpTxn txn = (TSHttpTxn)edata;
+  int status    = TS_ERROR;
+
+  switch (event) {
+  case TS_EVENT_HTTP_CACHE_LOOKUP_COMPLETE:
+    if (TSHttpTxnCacheLookupStatusGet(txn, &status) == TS_SUCCESS) {
+      if (status == TS_CACHE_LOOKUP_HIT_FRESH) {
+        PluginState *const pstate = static_cast<PluginState *>(TSContDataGet(cont));
+        // atomcally grab a handle
+        using InvHandle        = std::shared_ptr<std::vector<Invalidate>>;
+        InvHandle const vecinv = std::atomic_load(&(pstate->invalidate_vec));
+
+        if (vecinv) {
+          time_t const date    = get_date_from_cached_hdr(txn);
+          time_t const timenow = time(nullptr);
+          char *url            = nullptr;
+          int url_len          = 0;
+
+          for (Invalidate const &inv : *vecinv) {
+            if (date <= inv.epoch && timenow <= inv.expiry) {
+              if (nullptr == url) {
+                url = TSHttpTxnEffectiveUrlStringGet(txn, &url_len);
+              }
+
+              if (nullptr == url || 0 == url_len) {
+                break;
+              } else if (inv.matches(url, url_len)) {
+                TSHttpTxnCacheLookupStatusSet(txn, TS_CACHE_LOOKUP_HIT_STALE);
+                DEBUG_LOG("Forced revalidate - %.*s", url_len, url);
+                break;
+              }
+            }
+          }
+
+          if (nullptr != url) {
+            TSfree(url);
+          }
+        }
+      }
+    }
+    break;
+  case TS_EVENT_HTTP_TXN_CLOSE: // never called by global
+    TSContDestroy(cont);
+    break;
+  default:
+    break;
+  }
+
+  TSHttpTxnReenable(txn, TS_EVENT_HTTP_CONTINUE);
+  return 0;
+}
+
+} // namespace
+
+namespace
+{
+void
+setup_memory_allocation()
+{
+  if (&ts_malloc != pcre_malloc) {
+    pcre_malloc = &ts_malloc;
+    pcre_free   = &ts_free;
+  }
+}
+
+void
+setup_config_cont(PluginState *const pstate)
+{
+  if (nullptr == config_cont) {
+    DEBUG_LOG("creating config continuation");
+
+    // create reusable configuration file regex
+    constexpr char const *const expr = "^([^#].+?)\\s+(\\d+)\\s*$";
+    bool const cstat                 = config_re.compile(expr);
+
+    if (!cstat) {
+      ERROR_LOG("setup_config_cont: Unable to compile config_re");
+      return;
+    }
+
+    // set up the config continuation
+    config_cont = TSContCreate(config_handler, TSMutexCreate());
+    TSContDataSet(config_cont, static_cast<void *>(pstate));
+
+    // constantly run this for rule grooming
+    TSContScheduleOnPool(config_cont, CONFIG_TMOUT, TS_THREAD_POOL_TASK);
```

Comment:
It seems convoluted to schedule this, then ignore the event it generates, when timed_reload is false.

Reply:
convoluted is correct.  Even with timed_reload disabled any expired rules will still be pruned out.